### ISSUE TITLE
Shader compiler should not allow defining uniforms in local scope (as it's invalid)

### ISSUE THREAD

**Issue by LinuxUserGD** [godotengine/godot#15133]
**Godot version:**
latest master branch

**OS/device including version:**
Ubuntu Linux 16.04 64 bit

**Issue description:**
I've tried to port this shader (https://www.youtube.com/watch?v=RLAG4RbT-5U) to GD 3.0 and I get some errors in console while no line  is marked in the editor. This makes it impossible for me to code a 3d spatial fragment shader.

**Steps to reproduce:**

Code:
```

shader_type spatial;
void fragment() {
	sampler2D grass;
	sampler2D rock;
	sampler2D dirt;
	sampler2D splatmap;
	vec3 result;
	vec3 grasscol;
	vec3 rockcol;
	vec3 dirtcol;
	float grassval = texture(splatmap,UV).g;
	float rockval = texture(splatmap,UV).r;
	float dirtval = texture(splatmap,UV).b;
	grasscol = texture(grass,UV).rgb * grassval;
	rockcol = texture(rock,UV).rgb * rockval;
	dirtcol = texture(dirt,UV).rgb * dirtval;
	result = grasscol+rockcol+dirtcol;
}

```

Error Output:
```
drivers/gles3/shader_gles3.cpp:167 - SceneShaderGLES3: Fragment Program Compilation Failed: 0:1163(2): error: opaque variables must be declared uniform 0:1164(2): error: opaque variables must be declared uniform 0:1165(2): error: opaque variables must be declared uniform 0:1166(2): error: opaque variables must be declared uniform 0:1171(27): warning:m_splatmap' used uninitialized
0:1172(26): warning: m_splatmap' used uninitialized 0:1173(26): warning:m_splatmap' used uninitialized
0:1174(22): warning: m_grass' used uninitialized 0:1175(21): warning:m_rock' used uninitialized
0:1176(21): warning: m_dirt' used uninitialized drivers/gles3/shader_gles3.cpp:440 - SceneShaderGLES3: Fragment Program Compilation Failed: 0:1163(2): error: opaque variables must be declared uniform 0:1164(2): error: opaque variables must be declared uniform 0:1165(2): error: opaque variables must be declared uniform 0:1166(2): error: opaque variables must be declared uniform 0:1171(27): warning:m_splatmap' used uninitialized
0:1172(26): warning: m_splatmap' used uninitialized 0:1173(26): warning:m_splatmap' used uninitialized
0:1174(22): warning: m_grass' used uninitialized 0:1175(21): warning:m_rock' used uninitialized
0:1176(21): warning: m_dirt' used uninitialized drivers/gles3/shader_gles3.cpp:448 - Method/Function Failed, returning: __null drivers/gles3/shader_gles3.cpp:122 - Condition ' !version ' is true. returned: false ./drivers/gles3/shader_gles3.h:376 - Condition ' !version ' is true. returned: -1 ./drivers/gles3/shader_gles3.h:376 - Condition ' !version ' is true. returned: -1 ./drivers/gles3/shader_gles3.h:376 - Condition ' !version ' is true. returned: -1 ./drivers/gles3/shader_gles3.h:376 - Condition ' !version ' is true. returned: -1 ./drivers/gles3/shader_gles3.h:376 - Condition ' !version ' is true. returned: -1
```

**Minimal reproduction project:**
[splatmap.zip](https://github.com/godotengine/godot/files/1592132/splatmap.zip)

- [ ] I searched the existing [GitHub issues](https://github.com/godotengine/godot/issues?utf8=%E2%9C%93&q=is%3Aissue+) for potential duplicates.
**Similar issue:**  https://github.com/godotengine/godot/issues/13022

**Background information:**
```
shader_type spatial;
void fragment() {
	uniform sampler2D grass;
	uniform sampler2D rock;
	uniform sampler2D dirt;
	uniform sampler2D splatmap;
	vec3 result;
	vec3 grasscol;
	vec3 rockcol;
	vec3 dirtcol;
	float grassval = texture(splatmap,UV).g;
	float rockval = texture(splatmap,UV).r;
	float dirtval = texture(splatmap,UV).b;
	grasscol = texture(grass,UV).rgb * grassval;
	rockcol = texture(rock,UV).rgb * rockval;
	dirtcol = texture(dirt,UV).rgb * dirtval;
	result = grasscol+rockcol+dirtcol;
}
```
Error Output:
```
:4 - Expected expression, found: OUT drivers/gles3/rasterizer_storage_gles3.cpp:1671 - Condition ' err != OK ' is true.
```
The first line 'uniform sampler2D grass;' is marked and I got the same error for every hint (vec4, int, float, sampler2D). That's why I've removed "uniform".

**Comment by brainsick** [comment]
Move your uniforms out of the fragment() method.  The [documentation](http://docs.godotengine.org/en/latest/learning/features/shading/shading_language.html#uniforms) indicates that these are global to the shader, so I would define them as such.

**Comment by LinuxUserGD** [comment]
Thank you for your advice and work, thought that it was some engine bug. I close this issue now as the error was triggered by an invalid shader script.

**Comment by reduz** [comment]
This is a bug, I'm surprised the compiler allowed it

**Comment by akien-mga** [comment]
We have now entered release freeze for Godot 3.0 and want to focus only on release critical issues for that milestone. Therefore, we're moving this issue to the 3.1 milestone, though a fix may be made available for a 3.0.x maintenance release after it has been tested in the master branch during 3.1 development. If you consider that this issue is critical enough to warrant blocking the 3.0 release until fixed, please comment so that we can assess it more in-depth.
